### PR TITLE
add script to publish runtime-builders to GCS

### PR DESCRIPTION
**Pull request by nkubala**
This will be the heart of a job that will monitor the configuration files for the runtime builders, resolve their tags into pinned SHA256 digests and publish them to GCS.

Tests to come.

@sharifelgamal @dlorenc @znewman01

### CODE REVIEW

File: runtime_builders/publish_runtime.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--infile', '-i', help='file(s) to read')
+    parser.add_argument('--bucket', '-b',
+                        help='GCS bucket to publish runtime to',
+                        default='runtime-builders')
+    parser.add_argument('--language', '-l',
+                        help='the language associated with this builder')
+    args = parser.parse_args()
+
+    if args.infile is None:
```
Comment (dlorenc) [inline runtime_builders/publish_runtime.py]: I think you can mark these flags as required and then you can drop the error checking here.

File: runtime_builders/publish_runtime.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    if ':' not in image:
+        logging.info('Image does not contain tag: defaulting to "latest"')
+        base_image = image
+        target_tag = 'latest'
+    else:
+        base_image = image.split(':')[0]
+        target_tag = image.split(':')[1].strip('${}')
+
+    command = ['gcloud', 'beta', 'container', 'images',
+               'list-tags', base_image, '--format=json']
```
Comment (dlorenc) [inline runtime_builders/publish_runtime.py]: I think this will be a little easier with "describe" instead of list-tags.

Comment (nkubala) [inline runtime_builders/publish_runtime.py]: ah you're right, saves me the loop.

File: runtime_builders/publish_runtime.py (hunk 1 above)
Comment (nkubala) [inline runtime_builders/publish_runtime.py]: good call.

File: runtime_builders/publish_runtime.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+def _resolve_tag(image):
+    """
+    Given a path to a tagged Docker image in GCR, replace the tag with its
+    corresponding sha256 digest.
+    """
+    if image is None:
+        raise Exception('Please provide image')
+
+    if ':' not in image:
```
Comment (dlorenc) [inline runtime_builders/publish_runtime.py]: I think we might want to force people to be explicit here, and disallow implicit "latest". WDYT?

File: runtime_builders/publish_runtime.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        logging.info('Image does not contain tag: defaulting to "latest"')
+        base_image = image
+        target_tag = 'latest'
+    else:
+        base_image = image.split(':')[0]
+        target_tag = image.split(':')[1].strip('${}')
```
Comment (dlorenc) [inline runtime_builders/publish_runtime.py]: It might be more robust if we do this with a regular expression, that way we can fail if they have something unclear.

We can also use the same regular expression as a pre-submit test for the file.

File: runtime_builders/publish_runtime.py (hunk 3 above)
Comment (sharifelgamal) [inline runtime_builders/publish_runtime.py]: +1

Comment (nkubala) [inline runtime_builders/publish_runtime.py]: yeah this was something I was thinking about. I think that should be fine, we'll just need to make sure people are aware of it when they start writing their builders.

File: runtime_builders/publish_runtime.py (hunk 4 above)
Comment (nkubala) [inline runtime_builders/publish_runtime.py]: done.

File: runtime_builders/publish_runtime.py (hunk 2 above)
Comment (nkubala) [inline runtime_builders/publish_runtime.py]: update: `describe` requires that the active account has read permissions to the project associated with the image. because we can't guarantee this at runtime, going to go back to using `list-tags`.

File: runtime_builders/publish_runtime.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+]
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--infile', '-i', help='file(s) to read')
+    parser.add_argument('--bucket', '-b',
+                        help='GCS bucket to publish runtime to',
+                        default='runtime-builders')
+    parser.add_argument('--language', '-l',
```
Comment (znewman01) [inline runtime_builders/publish_runtime.py]: s/language/runtime-name/ ?

We may have multiple runtimes / language, it would appear.

Or would this publish *all* runtimes for a language? In which case I still don't think language is a great name.

File: runtime_builders/publish_runtime.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    'golang',
+    'dotnet',
+    'php',
+    'test'
+]
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--infile', '-i', help='file(s) to read')
```
Comment (znewman01) [inline runtime_builders/publish_runtime.py]: I don't think that help is very helpful...What should the files be? What do they indicate?

In addition to the fact that this represents a templated input file, I think the following would be really helpful:
- a short example file
- a brief description of the templating language
- a mention that multiple templating passes will happen

File: runtime_builders/publish_runtime.py (hunk 5 above)
Comment (znewman01) [inline runtime_builders/publish_runtime.py]: add choices=LANGUAGES ?

File: runtime_builders/publish_runtime.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+            fd, ofile = tempfile.mkstemp()
+            with open(ofile, 'w') as outfile:
+                outfile.write(s)
+            outfile.close()
+            os.close(fd)
+            return ofile
+
+        except yaml.YAMLError as e:
+            print(e)
+        finally:
```
Comment (znewman01) [inline runtime_builders/publish_runtime.py]: Doesn't the `with open(...)` do this?

File: runtime_builders/publish_runtime.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+            s = yaml.round_trip_dump(config)
+
+            fd, ofile = tempfile.mkstemp()
+            with open(ofile, 'w') as outfile:
+                outfile.write(s)
+            outfile.close()
+            os.close(fd)
+            return ofile
+
+        except yaml.YAMLError as e:
```
Comment (znewman01) [inline runtime_builders/publish_runtime.py]: Shouldn't you still raise in this case?

File: runtime_builders/publish_runtime.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+                image = step.get('name')
+                templated_step = _resolve_tag(image)
+                step['name'] = templated_step
+
+            s = yaml.round_trip_dump(config)
+
+            fd, ofile = tempfile.mkstemp()
+            with open(ofile, 'w') as outfile:
+                outfile.write(s)
+            outfile.close()
```
Comment (znewman01) [inline runtime_builders/publish_runtime.py]: `with ...` closes the file

File: runtime_builders/publish_runtime.py [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+            infile.close()
+
+
+'''
+Given a path to a tagged Docker image in GCR, replace the tag with its
+corresponding sha256 digest. 
+'''
+def _resolve_tag(image):
+    if image is None:
+        raise Exception('Please provide image')
```
Comment (znewman01) [inline runtime_builders/publish_runtime.py]: Prefer an exception like ValueError()

File: runtime_builders/publish_runtime.py [hunk 11]
```diff
@@ -19,6 +19,7 @@
 import json
 import logging
 import os
+import regex
```
Comment (znewman01) [inline runtime_builders/publish_runtime.py]: What's the reason to use this over the std library "re"?

File: runtime_builders/publish_runtime.py [hunk 12]
```diff
@@ -103,29 +106,34 @@ def _resolve_tag(image):
...
     if ':' not in image:
-        logging.info('Image does not contain tag: defaulting to "latest"')
-        base_image = image
-        target_tag = 'latest'
+        logging.error('Image must contain explicit tag!')
+        sys.exit(1)
     else:
-        base_image = image.split(':')[0]
-        target_tag = image.split(':')[1].strip('${}')
+        m = regex.search(TAG_REGEX, image)
```
Comment (znewman01) [inline runtime_builders/publish_runtime.py]: Shouldn't this be match(), not search()? It should have to be at the beginning, right?

File: runtime_builders/publish_runtime.py [hunk 13]
```diff
@@ -103,29 +106,34 @@ def _resolve_tag(image):
...
     corresponding sha256 digest.
     """
-    if image is None:
-        raise Exception('Please provide image')
-
     if ':' not in image:
-        logging.info('Image does not contain tag: defaulting to "latest"')
-        base_image = image
-        target_tag = 'latest'
+        logging.error('Image must contain explicit tag!')
```
Comment (znewman01) [inline runtime_builders/publish_runtime.py]: Why this change?

**Comment by znewman01** [comment]
This makes sense from a standalone perspective, but I'm not certain how it'll fit into the larger workflow.

I'm also concerned about how local development will work.

(These are just me rambling, you don't have to address them now)

File: runtime_builders/publish_runtime.py (hunk 11 above)
Comment (nkubala) [inline runtime_builders/publish_runtime.py]: the std library `re` doesn't support variable length lookbehind, while this one does. the regex that I'm using to parse out the tags from image names wouldn't work in the std library.

File: runtime_builders/publish_runtime.py (hunk 6 above)
Comment (nkubala) [inline runtime_builders/publish_runtime.py]: yeah good call, sorry I definitely neglected documentation here. I created a README that has basic explanations of everything going on here, with an example.

what do you mean by multiple templating passes? the script should only go through a file once and do the substitutions as it encounters the tags.

File: runtime_builders/publish_runtime.py (hunk 5 above)
Comment (nkubala) [inline runtime_builders/publish_runtime.py]: yeah, sorry this is pretty unclear. basically, since this script just takes in a bunch of templated cloudbuild files and then writes them to GCS, it's agnostic to which language runtime each file is associated with, so to make things more explicit I decided to have each run be parameterized with the associated runtime. so a sample run would look like `./publish_runtime.py --infile=cloudbuild.yaml.in --language=python`. but I agree that `runtime-name` is better

Comment (nkubala) [inline runtime_builders/publish_runtime.py]: didn't know you could do that :)

File: runtime_builders/publish_runtime.py (hunk 9 above)
Comment (nkubala) [inline runtime_builders/publish_runtime.py]: oh duh

File: runtime_builders/publish_runtime.py (hunk 8 above)
Comment (nkubala) [inline runtime_builders/publish_runtime.py]: ah yeah good call, I just wanted to wrap the error in something a little prettier, but I still want to `sys.exit` here

File: runtime_builders/publish_runtime.py (hunk 7 above)
Comment (nkubala) [inline runtime_builders/publish_runtime.py]: sure does, fixed

File: runtime_builders/publish_runtime.py (hunk 13 above)
Comment (nkubala) [inline runtime_builders/publish_runtime.py]: we decided that instead of defaulting to `latest`, we should require that all images used in build steps are qualified with a tag to eliminate confusion.

File: runtime_builders/publish_runtime.py (hunk 12 above)
Comment (nkubala) [inline runtime_builders/publish_runtime.py]: because I'm using a positive lookbehind when searching for the target string, it actually doesn't work if I try and do a `match` from the beginning of the string. I probably could have formatted this regex better haha.

**Comment by nkubala** [comment]
> This makes sense from a standalone perspective, but I'm not certain how it'll fit into the larger workflow.

the way I was envisioning it was: runtime maintainers make changes to their builder, and every time they do this script will publish a new fully qualified builder config to GCS. otherwise, anytime anyone is trying to kick off a build, they just pull a config file from GCS and run a container build as normal.

currently these are just being published as <runtime>-<datetime>.yaml, but I don't see any reason why we can't allow maintainers to also specify some name or tag that they can use to reference a specific builder at a later date. 

> I'm also concerned about how local development will work.

should be similar to other environments? just pull a builder config from GCS and run a container build as normal. if you're not using cloudbuild then this shouldn't apply.

File: runtime_builders/publish_runtime.py (hunk 12 above)
Comment (znewman01) [inline runtime_builders/publish_runtime.py]: Got it, thanks for the explanation.

File: runtime_builders/publish_runtime.py (hunk 13 above)
Comment (znewman01) [inline runtime_builders/publish_runtime.py]: Oh, and a user could still use e.g. latest to be explicit. Makes sense.

File: runtime_builders/publish_runtime.py (hunk 6 above)
Comment (znewman01) [inline runtime_builders/publish_runtime.py]: Argo has templating as well–so we fill in e.g. $PROJECT and $_OUTPUT_IMAGE. The templating format is different. It's a little mind-bendy

Comment (nkubala) [inline runtime_builders/publish_runtime.py]: i agree it is a little weird to have two different "types" of templating in one file. I could change the formatting of this template if you think they're too similar (for instance, using an `@` instead of a `$` for the delimiter).

**Comment by znewman01** [comment]
Up to you..I think as long as it's clear what will get subbed at each pass,
that's fine

**Comment by dlorenc** [comment]
We can also handle some of this with testing. For example, we can have a presubmit test that runs template expansion on the input file and then validates the container: fields all match the correct docker tag regular expression. That only works if we choose a template expansion syntax that *is not valid* as a docker image tag, though.

**Comment by nkubala** [comment]
yeah I don't think that will be too difficult, from their docs:

`
A tag name must be valid ASCII and may contain lowercase and uppercase letters, digits, underscores, periods and dashes. A tag name may not start with a period or a dash and may contain a maximum of 128 characters.
`

as long as we put some symbol (like `$`) in there it will be fine.

my main concern is that it might get confused with what we're currently using for things like final image name substitution (e.g. `images: ['${IMAGE_NAME}']`). the regex knows how to tell the difference, but config file writers might not.

**Comment by dlorenc** [comment]
If the templating is confusing, we could just have users specify "tagged" docker images, and resolve the tags.

`container: gcr.io/foo/bar:latest`

would get auto-resolved.

We'd skip anything with a digest:

`container: gcr.io/foo/bar@sha256:abc123`

This would make local development a bit easier. WDYT?

**Comment by nkubala** [comment]
my only concern with that is that in that case it's hard to "version" the builders, since even though the tags might stay the same the images associated with them may change. nothing will be stopping anyone from using those config files in a cloudbuild directly (since they're legal Dockerfiles), so that might make it tough to have reproducible builds. do you think that'll be an issue?

**Comment by znewman01** [comment]
If I understand Dan correctly: you could swap the tags with digests at
release time, so you'd have (1) reproducible builds, (2) easy local dev,
and (3) easy testing pre-release against the current tags.

The more I think about it the more I like this strategy.

**Comment by nkubala** [comment]
cool, yeah after thinking about it again I agree. updated to expect actual tags/digests, and to replace all tags and leave digests untouched before pushing to GCR.

**Comment by dlorenc** [comment]
> The more I think about it the more I like this strategy.

Yea at first I thought it would be a "feature" that we couldn't accidentally use an un-resolved tag, but after some more thought I like this approach. We can ensure official releases don't contain tags other ways.

File: runtime_builders/publish_runtime.py [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+import os
+import regex
+from ruamel import yaml
+import subprocess
+import sys
+import tempfile
+
+from google.cloud import storage
+
+LANGUAGES = [
```
Comment (dlorenc) [inline runtime_builders/publish_runtime.py]: Do you think it's worth it to validate against this list? It might make adding new configurations more difficult.

File: runtime_builders/publish_runtime.py [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--infile', '-i',
+                        help='templated cloudbuild config file.',
+                        required=True)
+    parser.add_argument('--bucket', '-b',
+                        help='GCS bucket to publish runtime to',
+                        default='runtime-builders')
+    parser.add_argument('--language',
```
Comment (dlorenc) [inline runtime_builders/publish_runtime.py]: Naming this "runtime-name" might be more clear.

File: runtime_builders/publish_runtime.py [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+            logging.error('Error when parsing image tag!')
+            logging.error(e)
+            sys.exit(1)
+
+    command = ['gcloud', 'beta', 'container', 'images',
+               'list-tags', base_image, '--format=json']
+
+    try:
+        output = subprocess.check_output(command)
+        entries = json.loads(output)
```
Comment (dlorenc) [inline runtime_builders/publish_runtime.py]: Could you log the full output of the list command?

File: runtime_builders/publish_runtime.py [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+                        choices=LANGUAGES)
+    args = parser.parse_args()
+
+    if args.language not in LANGUAGES:
+        logging.error('Invalid language \'{0}\' specified! '
+                      'Options:'.format(args.language))
+        logging.error(LANGUAGES)
+        return 1
+
+    templated_file = _resolve_tags(args.infile)
```
Comment (dlorenc) [inline runtime_builders/publish_runtime.py]: Could you log the templated_file here?

File: runtime_builders/publish_runtime.py [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+            for step in config.get('steps'):
+                image = step.get('name')
+                templated_step = _resolve_tag(image)
+                step['name'] = templated_step
+
+            s = yaml.round_trip_dump(config)
+
+            fd, ofile = tempfile.mkstemp()
+            with open(ofile, 'w') as outfile:
+                outfile.write(s)
```
Comment (dlorenc) [inline runtime_builders/publish_runtime.py]: Hmm, do we need to actually write this to a file? Could we keep it in memory and upload it that way?

File: runtime_builders/publish_runtime.py (hunk 16 above)
Comment (nkubala) [inline runtime_builders/publish_runtime.py]: I can, but that can be a ton of info since it will display every image digest (even the ones without tags). might make more sense to just log the entries for tagged images?

File: runtime_builders/publish_runtime.py (hunk 18 above)
Comment (nkubala) [inline runtime_builders/publish_runtime.py]: nah we don't, I think I might have just been doing this for debugging purposes in the first place. removed

File: runtime_builders/publish_runtime.py (hunk 17 above)
Comment (nkubala) [inline runtime_builders/publish_runtime.py]: :+1:

File: runtime_builders/publish_runtime.py (hunk 15 above)
Comment (nkubala) [inline runtime_builders/publish_runtime.py]: agreed, see above comment

File: runtime_builders/publish_runtime.py (hunk 14 above)
Comment (nkubala) [inline runtime_builders/publish_runtime.py]: yeah, I really only did this for naming purposes (since the files that are passed in will all be called "cloudbuild.yaml.in" or something generic). gonna go ahead and remove this though in favor of passing a required "runtime-name" parameter.
